packages/libs/deer/src/lib.rs

```diff
@@ -6,4 +6,212 @@
 #![allow(clippy::module_name_repetitions)]
 #![forbid(unsafe_code)]
 
+use alloc::string::String;
+
+use error_stack::{IntoReport, Result};
+use num_traits::ToPrimitive;
+
+pub use crate::{error::Error, number::Number};
+
+mod error;
+mod number;
+
 extern crate alloc;
+
+/// Lookup of Type
+///
+/// This is used as the signal for the type that is currently present, [`Deserialize`]
+/// implementations may use this to special case certain values.
+#[derive(Debug, Copy, Clone)]
+pub enum PrimitiveType {
+    /// null type
+    Null,
+    /// bool type
+    Bool,
+    /// number type
+    Number,
+    /// string type
+    String,
+    /// array type
+    Array,
+    /// object type
+    Object,
+}
+
+macro_rules! derive_from_number {
+    ([$($name:ident; $method:ident),*]) => {
+        $(derive_from_number!(#internal, $name; $method);)*
+    };
+
+    (#internal, $name:ident; $method:ident) => {
+        /// Automatically implemented convenience method, which uses [`Self::number`] to extract
+        /// a value of the primitive type, will otherwise error out.
+        ///
+        /// # Errors
+        ///
+        /// Current value is either not a number or wasn't able to be casted to the primitive type
+        fn $name(self) -> Result<$name, Self::Error> {
+            self.number()?
+                .$method()
+                .ok_or_else(|| Self::Error::message("provided value too large or too small"))
+                .into_report()
+        }
+    };
+}
+
+/// A **data format** that can deserialize any data structure which is supported by deer.
+///
+/// This traits defines the deserialization half, while [`Deserialize`] is used to use the
+/// deserializer to convert that data into the instance of a type, or fail if that was not possible.
+///
+/// `deer`s focus is on simplicity, therefore the data model is a lot less specific than ones used
+/// by other crates (like [`serde`]).
```

Comment:
Is there a specific reason, why you chose this approach? Serde uses the visitor pattern to fallback to more generic primitives. IMO the primitives are too broad. For me it's a huge difference, if a number is a floating point or an integral. Also, it's different, if a string can be borrowed or not. Also, I don't like `null` == `None` and `()` (thinking of `null` vs `undefined` in javascript)

Reply:
For now, values are not going to be able to be borrowed. The intermediary `Number` struct covers the difference between integral and floating, which stores those (analogous to `serde_json`, `serde_yaml`, and `rmpv`).

serde uses the `deserialize_*` functions as hints for non-self-describing formats. In our use case, I don't see a benefit of visitor pattern vs. "direct" access.

The flexibility comes at the cost of more complexity, which I'd like to avoid.

Reply:
Regarding the `null == None` and `null == ()`. The idea here is not to encode anything Rust-specific. Instead, `null` encodes the absence of a value. How you interpret that absence of the value (`Option::None` vs. `()`) is up to the `Deserialize` implementation.

`Option<T>` would `visit_null()` to `Option::None`, while all other would delegate to `T`, in the case of `()` only `visit_null` would be valid.

Reply:
new approach in [c8a8e45](https://github.com/hashintel/hash/pull/1170/commits/c8a8e452b4d72c95b18b136f85172f35768d77b8)

Reply:
Great, thanks for implementing it 🙂 

With the visitor pattern in place, having very broad primitives is fine I think as they can be extended very easily later on, so I'm resolving this.